File: figures/08-graphs/snippet-dfs-iterative.py
def dfs_iterative(g, s):
    discovered = [False] * g.vertices()
    stack = [s]
    discovered[s] = True
    visit_order = []
    while stack:
        u = stack.pop()
        visit_order.append(u)
        for v, _ in reversed(g.edges_from(u)):
            if not discovered[v]:
                discovered[v] = True
                stack.append(v)
    return visit_order

File: figures/08-graphs/snippet-dfs-iterative.java
    public static int[] dfsIterative(Graph g, int s) {
        int n = g.vertices();
        boolean[] discovered = new boolean[n];
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(s); discovered[s] = true;
        List<Integer> visitOrder = new ArrayList<>();
        while (!stack.isEmpty()) {
            int u = stack.pop();
            visitOrder.add(u);
            List<Graph.Edge> neighbors = g.edgesFrom(u);
            for (int i = neighbors.size() - 1; i >= 0; i--) {
                int v = neighbors.get(i).to();
                if (!discovered[v]) { discovered[v] = true; stack.push(v); }
            }
        }
        return visitOrder.stream().mapToInt(Integer::intValue).toArray();
    }
